### DISCUSSION TITLE
SD card error and blank screen

### DISCUSSION THREAD

**Discussion by HauptlingHD** [sanni/cartreader#842] [Need help]
Hi

I've just finished soldering the main PCB of the cartreader. I first flashed the latest 12.8 release and wanted to turn on the cartreader, but nothing really happened. The screen of the MKS MINI stayed blank. The RGB LED does change color when I turn the knob. So I searched around and learned that with this problem, it is advised to check the output in SERIAL_MODE. After getting an SD card error, I checked for continuity from the 2560 to the EXP connector, and everything seems fine. On one pin, I also get continuity to the SD card slot. I believe this is also correct. So I searched through the discussions and found many people having problems with certain SD card models and formatting.

I tried older and newer SanDisk models ranging from 1 to 32 GB and even a micro SD with an adapter. All were tested with formatting by the SD formatter and exFAT formatted with Windows Explorer. Always SD card error on the serial output...

Does someone have a suggestion on what to check next?

PS: All SD cards work on all other devices and are tested sector by sector.

**Comment by Ancyker** [comment]
SD cards 2GB and under should be formatted FAT16, cards 32GB and smaller should be formatted with FAT32, and larger than 32GB can be FAT32 or exFAT.

To properly format an SD card and remove all doubt:

1. Press `Win` + `R`.
2. Type `cmd`.
3. Press `Ctrl` + `Shift` + `Enter`.
4. Click "Yes" if prompted.
5. Type `diskpart` and press `Enter`.
6. Type `list disk` and press `Enter`.
7. Figure out which disk is the SD card. If you aren't sure then remove it and run the above command again, reinsert it, and run the command again. The disk that vanished and reappeared is the SD card. If in doubt, disconnect all other drives. `C:` should always be `Disk 0`.
8. Once you are sure you've found which disk is the SD card, type `select disk N` where N is the number of the disk (i.e. `select disk 1`) and press `Enter`.
9. Type `detail disk` and press `Enter`.
10. Ensure the information looks reasonable (likely it will mention being an SD USB reader, i.e. mine says "SABRENT SD USB Device").
11. Type `clean` and press `Enter`.
12. Type `convert basic` and press `Enter`. This command may fail if the disk is already a basic disk, ignore it.
13. Type `convert mbr` and press `Enter`.
14. Type `list partition` and press `Enter`, it should say there are no partitions on the disk.
15. Type `create partition primary` and press `Enter`.
16. Type `list partition` and press `Enter`.
17. Verify partition 1 exists, that its type is primary, and that it is the only partition.
18. Type `select partition 1` and press `Enter`.
19. Type `active` and press `Enter`.
20. Type `format fs=fat32 label="OSCR" quick` and press `Enter`.
21. Type `exit` and press `Enter`.
22. Type `exit` and press `Enter` (again) to close the Command Prompt.

You should now have a clean disk:
![image](https://github.com/sanni/cartreader/assets/7105333/62c23a78-d23e-400b-a834-2a8f613b9954)

If you are still having issues let me know :)

Example output:
```
C:\Windows\system32>diskpart

Microsoft DiskPart version 10.0.19041.964

Copyright (C) Microsoft Corporation.
On computer: AQUILA

DISKPART> list disk

  Disk ###  Status         Size     Free     Dyn  Gpt
  --------  -------------  -------  -------  ---  ---
  Disk 0    Online         5723 GB  1024 KB        *
  Disk 1    Online           29 GB      0 B

DISKPART> select disk 1

Disk 1 is now the selected disk.

DISKPART> detail disk

SABRENT SD USB Device
Disk ID: DBD446A5
Type   : USB
Status : Online
Path   : 0
Target : 0
LUN ID : 0
Location Path : UNAVAILABLE
Current Read-only State : No
Read-only  : No
Boot Disk  : No
Pagefile Disk  : No
Hibernation File Disk  : No
Crashdump Disk  : No
Clustered Disk  : No

  Volume ###  Ltr  Label        Fs     Type        Size     Status     Info
  ----------  ---  -----------  -----  ----------  -------  ---------  --------
  Volume 3     F   OSCR         FAT32  Removable     29 GB  Healthy

DISKPART> clean

DiskPart succeeded in cleaning the disk.

DISKPART> convert basic

The selected disk is not a dynamic disk.
Select a dynamic disk to convert to basic.

DISKPART> convert mbr

DiskPart successfully converted the selected disk to MBR format.

DISKPART> list partition

There are no partitions on this disk to show.

DISKPART> create partition primary

DiskPart succeeded in creating the specified partition.

DISKPART> list partition

  Partition ###  Type              Size     Offset
  -------------  ----------------  -------  -------
* Partition 1    Primary             29 GB  1024 KB

DISKPART> select partition 1

Partition 1 is now the selected partition.

DISKPART> active

DiskPart marked the current partition as active.

DISKPART> format fs=fat32 label="OSCR" quick

  100 percent completed

DiskPart successfully formatted the volume.

DISKPART> exit

Leaving DiskPart...

C:\Windows\system32>
```

**Reply by HauptlingHD** [reply]
Thank you for the detailed answer, but the error is still there. Isn't there some way of getting a clearer error description? What does "SD Error" mean exactly? Can I bypass the error and run the serial diagnostic program over serial? It says "check pins D50-53" in the wiki when experiencing this error. Is it sufficient to test for continuity from the 2560 to the EXP connectors, or is there something else to check?

**Reply by HauptlingHD** [reply]
I meant the SELF diagnostic program.

**Reply by cmcm711** [reply]
Are you copying the files in the "SD card" folder to the SD card after formatting?

**Reply by HauptlingHD** [reply]
Yes, I copy the contents of the SD folder to the root of the SD card and mark them as hidden, as described by the guide.

**Comment by Ancyker** [comment]
Sounds like you have either a bad solder connection to the SD card or the SD card reader is bad. Can you take a picture of the back of the LCD module?

**Comment by HauptlingHD** [comment]
![PXL_20230831_190312033 - Kopie](https://github.com/sanni/cartreader/assets/67075707/b1063ec5-6a65-4fe3-9156-63a06e9e485e)
![PXL_20230831_190317541 - Kopie](https://github.com/sanni/cartreader/assets/67075707/cdb169b9-5d1e-4ea2-906a-0f9dfff0908d)
![PXL_20230831_190322089 - Kopie](https://github.com/sanni/cartreader/assets/67075707/83505ad1-0660-4c5d-a2ba-7a5afd42cc40)
![PXL_20230831_190343346 - Kopie](https://github.com/sanni/cartreader/assets/67075707/2ce4bdb0-952e-4187-996b-87354ab4e727)
![PXL_20230831_190122508 - Kopie](https://github.com/sanni/cartreader/assets/67075707/f7b93222-71b3-4ab4-8f2a-f8177b630a3f)
![PXL_20230831_190304676 - Kopie](https://github.com/sanni/cartreader/assets/67075707/ce69276c-3722-43ea-a209-d20569e5f1f6)


There you go

**Reply by Ancyker** [reply]
With the OSCR unplugged and the LCD disconnected, test for continuity from the pins on the Arduino (not the main PCB) and the cable plugged into the leftmost LCD header. You can use a spare header connected to the cable to get easier access to the pins. If D50-53 all have continuity then test resistance, it should read as nearly zero. Here is a pin map:

![image](https://github.com/sanni/cartreader/assets/7105333/0523e7ff-e676-4d34-a35d-73ad4054250e)

If this all checks out, then the issue is likely the module itself. 

Edit: Note: Don't stop checking pins when you verify. Check ALL the pins on one side. I recommend putting a probe on the Arduino and using the other to check all of the pins on the cable. Do this for every pin on the mega. If you only test pin-to-pin you won't find shorts.

**Reply by HauptlingHD** [reply]
I tested everything. From the Arduino to the main PCB connector, there are 0 ohms and no shorts. Then I measured everything with the cable. I measured 2 ohms on all pins. I also measured the cable on its own by pressing my probes against the open cable ends that are sticking out of the connector. Only the conductor of the cable measures 1,5 Ohm. With spare connectors on each end, it measures 2 ohms. For reference, I measured an old IDE cable I had lying around, and this much longer cable only has 1 ohm of resistance. Are 2 ohms too high? While waiting for the answer, I will also make a new short cable out of the IDE cable and test if this works.

**Reply by HauptlingHD** [reply]
The cable did not make a difference, although the resistance was much lower. I also have a new problem now. When I tried to reflash the Arduino, I got "avrdude: stk500v2_ReceiveMessage(): timeout" in the IDE. It doesn't want to flash anymore and there is no output on the serial monitor.🙃

**Reply by Ancyker** [reply]
Does it work without the LCD connected? Make sure to have it set to 5V.

**Reply by HauptlingHD** [reply]
I just get

avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_ReceiveMessage(): timeout
avrdude: stk500v2_getsync(): timeout communicating with programmer

No matter what is connected. Yes, I am using the 5V setting.

**Reply by Ancyker** [reply]
Does the light on the Arduino come on when USB is connected?

**Reply by HauptlingHD** [reply]
Yes, the red and green LEDs turn on, and Windows recognizes USB-SERIAL CH340 (COM3). I literally didn't do anything but try to reflash with the #define HW5 option instead of #define SERIAL_MONITOR.

**Reply by HauptlingHD** [reply]
The latest Arduino IDE gives the same error, even though it automatically detects the COM port and model.

**Reply by HauptlingHD** [reply]
I let the Arduino connect to my PC for a while, and the red LED blinked twice, but the error is still there.

**Reply by Ancyker** [reply]
It sounds like the reflash failed and the bootloader is corrupt. You might have to burn a new bootloader. This requires a second Arduino (pretty much any, even an Uno works). As you're the second person in the last few days to end up with a corrupt bootloader I should write a guide on it. 

In the meantime, if you have a 2nd Arduino, you can use the instructions in my comment on #619 to burn a new bootloader. You can skip editing the `.ini` step as Sanni has started including the fix with the releases. Start from where it says "Download this file" (you don't need to download the file, though).

**Reply by HauptlingHD** [reply]
I flashed the fixed bootloader using an Arduino UNO, and now I am able to flash the 2560 again. Now I am back to the initial issue.

```
怘⸮f⸮⸮~f⸮`⸮xf⸮fxxf⸮⸮f⸮⸮⸮
Cartridge Reader
2023 github.com/sanni
SD Error

Press Button...
```

Is there a way to verify that the display is working outside of the OSCR firmware? I am not much of a C++ programmer. Even if I find an example sketch or something similar, I am not sure if the pins that are used would be the same as with the OSCR. Or do you see no point in trying that out? I do not see how, but if  you suspect the MKS MINI to be defective, then I guess I would have to try to get it refunded and order a new one. Maybe it was DOA, but I am not sure how to verify that it is.

**Reply by HauptlingHD** [reply]
Also, it could be a huge help if someone could send a known-good SD card that worked on another OSCR before to a German address. That way, I could be sure it was not a problem with my SD cards, since the error "SD Error" would lead me to believe so. If anyone is interested, let me know.

Alternatively, it would be nice to know exactly what SD cards are known to work. If there is a model that I can buy that works for someone else, I can test it and be sure that it is my MKS MINI that is the problem.

**Reply by Ancyker** [reply]
Does the OSCR stay powered when set to 3.3V?

**Reply by HauptlingHD** [reply]
Yes, the soldered wire on the Arduino and the bridged connection on the MKS MINI seam to work. The OSCR behaves the same in 5 and 3.3 V mode.

**Reply by HauptlingHD** [reply]
I tested two newly bought SD cards from Samsung and SanDisk. No luck. :( I guess my MKS MINI is bed. Just one last thing before I order a new one. Are the Pins D50-53 supposed to have a direct connection to the SD card itself, or does the Arduino "talk" to a chip on the MKS MINI? I'm getting continuity for ground and between pin D50 of the Arduino and the X/DATA1 pin of the SD reader. (referencing a picture of the "SD on back" model) Can someone confirm that this is correct?
![For-Makerbase-MKS-MINI12864-V3-Insert-SD-Card-Front-Side-LCD-Smart-Display-Screen-RGB-Backlight jpg_640x640](https://github.com/sanni/cartreader/assets/67075707/da71bb25-1419-42b9-a5d8-79f76ef5d3ec)

**Reply by Ancyker** [reply]
The SD card slot is connected through the chip above EXP2 which converts the 5V logic to 3.3V.

**Comment by HauptlingHD** [comment]
I ordered a new MKS MINI and now the SD card works.🥳 Unfortunately, now I have a similar problem. In "normal" mode (#define HW5), the OSCR behaves similarly to before. The backlight of the LCD turns on, and the LCD stays blank. Before, the RGB LED of the button changed color when turning, now it does not. In serial monitor mode, I get the same SD Error as before with no SD inserted (obviously), but when I insert an SD card, the serial output gets stuck at some point. Sometimes it writes:`怘⸮f⸮⸮~f⸮`⸮xf⸮fxxf⸮⸮f⸮⸮⸮
Cartridge Reader
2`

or`Cartridge Reader
2023 github.com/sanni

OS`

**Reply by HauptlingHD** [reply]
Now it works 100% with all SD cards. There must have been dirt on the contacts of the reader or something.
